### ISSUE TITLE
Create new MacPorts file for wxMaxima

### ISSUE THREAD

**Issue by MSoegtropIMC** [wxMaxima-developers/wxmaxima#1220]
The most recent MacPorts file for wxMaxima is 18.02. I am working on a new Macports file but need some help debugging. This issue is mostly intended to discuss this, since the Macports file will be integrated at the MacPorts project.

https://trac.macports.org/ticket/60453

https://trac.macports.org/ticket/60457

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : I have a 19.12.4 version compiled and it basically works. The 20.X versions I can't install as yet, but it shouldn't be hard to fix this.

With 19.12.4 I get wxMaxima to start maxima (I see the process and its parent is wxMaxima) and I also see a port 4010, but there doesn't appear to be any communication between wxMaxima and Maxima. In the firewall wxMaxima is allowed to receive incomming local requests. How should I debug this

**Comment by gunterkoenigsmann** [comment]
I believe there was a security mechanism on MacOs that either disallows wxMaxima or maxima to talk to the other program by default somewhere. But I have no idea on how to tell MacOs that the two programs are allowed to do so

Will make a bugfix release of wxMaxima, presumably on Monday. Is there anything I can do about the 20.X problem?

**Comment by MSoegtropIMC** [comment]
The 20.X problem should just be a simple path problem.

I am not sure what the communication issue is - as I said I know where to set the I/O allowance and this should be fine. Are there any debug log options to wxMaxima?

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : I did some more tests and also added the old 18.02 version as variant to my macports file so that I can switch easily. The result is that 18.02 does work and 19.12.4 doesn't. It does find maxima and starts it with a reasonable command line (including -s 4010) but it doesn't talk to it. I am quite sure that the firewall settings are correct. So to me this looks more like an issue in wxMaxima. I once had a 19.03 version running - don't ask me how I did it - which did also work. I will add 19.03  to my ports file as well and see. And of cause I will also try to fix the 20.x install.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : I did a few more tests: 19.05.7 does work, 19.12.4 does not (no communication between wxMaxima and maxima). So it looks like this did break on Mac in between. Do you remember some changes you did in this time which might cause this or shall I do a bisection search?

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : sorry please ignore the statement about 19.05.7. Retesting it did not work. It might be that I still had the 18.X version running and it would just open that or the issue is random. Switching around between various versions for 10 minutes I can say that 18.02 works reproducibly and 19.X doesn't.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : ok, I did a bisection search of tags - it was pretty quick:

18.02.0 does work
18.10.0 and several 19.X versions I tested don't work

**Comment by MSoegtropIMC** [comment]
There are 514 commits between these 2 tags. Do you think it makes sense to do an intersection search (likely 9 steps)? Probably time to write a script ...

**Comment by gunterkoenigsmann** [comment]
If you can do such a thing that would be extremely helpful. As I don't own a Mac and it looks like there is nothing generally wrong with wxMaxima's code (it works on several wxWidgets versions on Linux and it seems to run fine on MS Windows, as well) I don't know where to search, otherwise.

**Comment by gunterkoenigsmann** [comment]
One thing I have found - and will now experimentally revert - is: At one point I have changed wxMaxima's communication port: It used to be 4100, which is in the range 1024...49151 that is reserved for ports assigned by the IETF. Not that if the communication port is changed back to 4100 in the configuration dialogue everything starts to work again...

**Comment by gunterkoenigsmann** [comment]
Didn't dare to revert it, but moved the standard port to 49152 instead.

**Comment by MSoegtropIMC** [comment]
Thanks a valuable hint - it might be that changing the port confuses the firewall when switching between versions. Anyway I thought I make a script which builds every nth commit from a list of commits and then activates one after another (with MacPorts you can have several versions compiled in parallel and then quickly switch between them which si quite convenient for this). This might be useful for debugging other stuff as well. I will do this over the weekend.

**Comment by gunterkoenigsmann** [comment]
That sounds like you could be interested in the command "git bisect": A command that first looks in which half of a range an error resides and then looks in which half of that range it will be in and so on. Using "git bisect" means that for finding an error in 200 commits at maximum 8 builds are required.

Today I have changed the following things hoping that one of them might somehow cause wxWidgets to avoid using the broken code path:

- Removed a flag to wxServer::wxServer() we don't really need
- Changed the default port (but the config overrides that setting on existent installs)
- Changed the way we detect how we address "localhost"

With much luck any of these errors makes wxMaxima work for MacOs again without somehow breaking MS Windows.

**Comment by MSoegtropIMC** [comment]
I did the bisection now. I didn't use git bisect because I wanted to have as little changes as possible to the MacPorts fetch/build infrastructure, but I will also give it a try. I am new to MacPorts and don't want to make to many changes in one step.

The result:
```
ab4154a3fef56267018fef66bb6f10d451ef2334 => OK
2c1b606467edeef9bff5f2e3fc152a805b875a9a => No connection
ab4154a3fef56267018fef66bb6f10d451ef2334 => OK
2c1b606467edeef9bff5f2e3fc152a805b875a9a => No connection
```
(I switched a few times to make sure it is reproducible).

As it looks the culprit are a few innocent outputs to stderr (the only difference between these 2 versions). I think what happens is that it hangs on the output to stderr because no console is connected and that's it then. I guess there is a separate thread for communication with Maxima since I can still type / edit, but as sonn as I hit Shift+Enter that's it.

I guess the best way forward is to switch to building with MacPorts from a local git repo, reproduce these results this way and then patch this and then rebase to various versions.

**Comment by MSoegtropIMC** [comment]
P.S.: I saw that the prints to stderr are removed in the tagged releases - they just mess up the bisection. I am not sure what cause the faults in the release. Unfortunately my MacPorts database got corrupted so I have to reinstall the things I need - will take a while.

There also seems to be some state kept somewhere I need to clean. There are some versions which work reproducibly and later they don't work reproducibly. Do you have a list of all config files I should delete between tests?

**Comment by MSoegtropIMC** [comment]
Some progress: some versions of wxMaxima create an invalid wxMaxima-init.mac file which keeps these and possibly other wxMaxima versions from connecting. If I delete wxMaxima-init.mac, I can run 18.10.0. The older 18.2.0 version shows some error in the wxMaxima window when the wxMaxima-init.mac file is invalid. Newer versions (like 18.10.0) simply don't connect to Maxima then. A guess of what might happen: maybe from some version on wxMaxima writes to stderr if the wxMaxima-init.mac file is invalid. The observable effect of writing to cerr in the code and of having an invalid wxMaxima-init.mac file is identical.

I am testing tagged versions beyond 18.10.0 now.

**Comment by gunterkoenigsmann** [comment]
That sounds like being a lot of progress, actually. If I understand it right

- The problem seems to lie in the wxMaxima-init.mac file, not in the code of wxMaxima
- And it is possible that the problem is that wxMaxima outputs something to stderr and that MacOs for some reason believes it has to disable the connection to maxima on this case.
 
With a bit of luck if you send me a broken wxMaxima-init.mac-File this file allows me to reproduce the output to std::cerr on my linux computer - and therefore to find out how to resolve the problem.

Thanks a lot for all of this work: Bisecting tends to be a very time-consuming business.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : I did some more tests / bisections. My status is as follows:

- the latest version which works is 20.01.2

- from 20.01.3 on the executable is missing from the installation. I don't get build errors, but the file `/Applications/MacPorts/wxMaxima.app//Contents/MacOS/wxmaxima` is missing. I have the same with all newer versions I tested, including `20.03.1` and `20.02.4`. Also the file `/Applications/MacPorts/wxMaxima.app//Contents/Info.plist` is missing. Otherwise the install folder contents is identical.

- I am not sure how wxMaxima is supposed to find gnuplot. I have a file `~/.maxima/wxmaxima-init.mac` with contents `gnuplot_command:"/opt/local/bin/gnuplot"$` but I don't know how this file comes to existence. Older versions of wxMaxima seem to create this file (some older versions create bogus files which leads to the hangs I described, I attached a bugs version of the file). When I delete the file and start wxMaxima, I would expect that the file is created, but it is not.

 - Newer versions have the Windows issue I described in (https://github.com/andrejv/wxmaxima/issues/6). For some versions the dangling pane is the history pane, for 20.01.2 it is the debug message pane. I have to enable the Debug Message pane, otherwise I get strange windows overlapping with / hiding the main window.

- In case I have a bad init file in `~/.maxima`, wxMaxima hangs on communication with maxima. Older version (18.02) display error messages in the main Windows about the bogus file. New versions don't, but hang. I expect that the issue is that newer version write the message to stderr, which also leads to hangs (which makes bisection hard, because some intermediate versions to write to stderr - tagged versions usually don't). This was my main issue with testing newer versions. Some old versions, I don't know which, seem to have created the bogus file. I attached a bogus and a good file. The difference is that in the bogus file the quotes are UTF-8 characters. It might help to figure out why writing to stderr leads to hangs and fix this.

I think if you explain me how the init file is supposed to be created, I can create a new MacPorts file for 20.01.2, explaining the situation in the notes (e.g. ask people to create the file manually). Possibly I can also create it from the MacPorts file, but I think this is not recommended. 

[InitFiles.zip](https://github.com/wxMaxima-developers/wxmaxima/files/4360712/InitFiles.zip)

**Comment by gunterkoenigsmann** [comment]
Normally during the build process make should output the info where it installs the binary to. At least it does do so in https://travis-ci.org/github/wxMaxima-developers/wxmaxima/jobs/664597997
Perhaps that gives us a hint on what went wrong during installation. The init file problem should be resolved in the current release. At least I hope it is; The gnuplot location if I remember the logic well wxMaxima asks maxima for. Don't know how maxima guesses where gnuplot is installed. though.

**Comment by tomio-arisaka** [comment]
The wxMaxima port file is made up of a Portfile and some patch files.
The "patch-src_Dirstructure.cpp.diff" file resolves the path to maxima in MacPorts.

If you execute "port note wxmaxima", then you can get the following note:

```
$ port note wxmaxima
wxMaxima has the following notes:
  Be careful: 
   If you want to use wxMaxima with default setting, 
   you should enter the command 
   "open -a /Applications/MacPorts/wxMaxima.app" 
   on Terminal.app in order to launch wxMaxima. 
   Because you must let wxMaxima know the path to software 
   installed with MacPorts. 
  
   On the other hand, 
   when you launch directly wxMaxima on the Finder, 
   the path to maxima (e.g. "/opt/local/bin/maxima") is automatically set 
   on the configuration dialog. 
   But you should enter the maxima expression 
   gnuplot_command:"/opt/local/bin/gnuplot"$ 
   in order to use gnuplot. 
  
   Some configuration files are automatically created by wxMaxima. 
   You can enter the command 
   "find $HOME/Library -name '*wxMaxima*'" 
   on Terminal.app in order to find them.
```
How to resolve the path to gnuplot is using AppleScript or shell-script.
See <https://sourceforge.net/projects/maxima/files/Maxima-MacOS/5.43.0-MacOSX/>

**Comment by MSoegtropIMC** [comment]
The path to maxima is fine without the patch in recent wxMaxima.

Also there are no tricks required to get around OSX gatekeeper. MacPorts handles this automatically (by setting a file system flag to its Application and binary folders).

What I wonder is how the path to gnuplot is created. @gunterkoenigsmann said wxMaxima gets it from Maxima, maxima apparently get's it from wxmaxima-init.mac and this file appears to be generated (sometimes) by wxMaxima. Of cause one can handle this with a script, but this should be automatic. wxMaxima should find gnuplot in the same way it finds maxima.

**Comment by MSoegtropIMC** [comment]
@tomio-arisaka : thanks for the link to the .dmg file. I must say I don't think that this is the best way of doing it. If wxMaxima needs a patch, a patch in a Macports file is only a temporary solution. One should always push such changes upstream. Luckily recent wxMaxima doesn't need the patch any more. Also the best way to publish a port file is to do a pull request to the MacPorts project. Working with local port files is intended for port file developers and not for distribution purposes. Finally if a port like wxMaxima has dependencies like maxima and in turn SBCL, these should be mentioned in the port file so that MacPorts installs them automatically.

The goal should always be that people just say `port install wxMaxima`. In case you are the author of the .dmg file, I would suggest that we maintain the wxMaxima MacPorts file together. It currently doesn't have a maintainer. Two people are better, one does changes and the other tests, reviews and approves it.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : In master I see these lines:

https://github.com/wxMaxima-developers/wxmaxima/blob/abae227abb9224d5a76257ecb3af202f1d8f197b/src/wxMaxima.cpp#L5789-L5808

https://github.com/wxMaxima-developers/wxmaxima/blob/abae227abb9224d5a76257ecb3af202f1d8f197b/src/wxMaxima.cpp#L3710-L3714

This looks a bit inconsistent. There should be only one way to find gnuplot.

Then I see:

https://github.com/wxMaxima-developers/wxmaxima/blob/abae227abb9224d5a76257ecb3af202f1d8f197b/src/ConfigDialogue.cpp#L1253-L1259

As far as I can tell the wxmaxima-init.mac file is only written when the config dialog is closed. Since the file seems to be vital, this is unfortunate. On the other hand it looks like wxMaxima gives the gnuplot path directly to maxima as well (in the second code snippet I posted above). I don't know if this kind of doubling is helpful. I think it is better to use one or the other method.

Btw.: If gnuplot is installed with MacPorts and MacPorts default settings, it ends up in `/opt/local/bin/gnuplot`. So in the search for gnuplot, wxMaxima should consider this path (I think it does but please double check).

I see what I can do about the missing executable file.

**Comment by MSoegtropIMC** [comment]
P.S.: If I delete the `~/.maxima/wxmaxima-init.mac` file and enter the preferences dialog, navigate to the startup commands tab and close the preferences dialog, an invalid startup file is written. It contains fancy UTF-8 double quotes `E2 80 9D` instead of normal ASCII quotes `22`. Is this Mac specific? This is tested on tag `20.01.2`.

**Comment by MSoegtropIMC** [comment]
One more question: which wxWidgets version to you prefer? The latest file in MacPorts uses 3.0 but has an option for 3.2 - something I would prefer to remove - it is just a source of inconsistent hard to reproduce behaviour.

**Comment by gunterkoenigsmann** [comment]
About wxWidgets: I hope that one of 3.0 and 3.2 works. On Linux 3.2 is way more stable, on MS Windows, too. But on the Mac I have no idea.

---

Mac computers are trying to be intelligent in many places. In which settings do the "smart quotes" appear? Perhaps there is a way to turn the intelligence off.

**Comment by gunterkoenigsmann** [comment]
About gnuplot: If you have an idea on how to improve the heuristics that searches for gnuplot I am happy. But as MacPorts, Fink, Homebrew and the DMG on the mac differ from each other in the idea on where to install stuff, as do ArchLinux, Ubuntu, SuSE, BSD and MS Windows we always will need a long search path.

---

The settings file should automatically be written once the wxConfig object is destroyed - which should happen automatically on closing the program. Since on the Mac only one instance of wxMaxima is running at a time that should be sufficient. At least in theory.

Many thanks in advance for helping to find out what has happened to the executable file.

**Comment by MSoegtropIMC** [comment]
> Mac computers are trying to be intelligent in many places. In which settings do the "smart quotes" appear? Perhaps there is a way to turn the intelligence off.

The start up commands tab in the settings dialog contains
```
gnuplot_command:”/opt/local/bin/gnuplot”$
```
As far as I can see already the text in the edit control contains the bad unicode quotes.

**Comment by MSoegtropIMC** [comment]
> I hope that one of 3.0 and 3.2 works. On Linux 3.2 is way more stable, on MS Windows, too. But on the Mac I have no idea.

As far as I can tell both work. I just wanted to say that I don't see a good reason for leaving this choice to the user. I will switch to 3.2 then. We can still think about this choice in case there are issues.

**Comment by MSoegtropIMC** [comment]
> The settings file should automatically be written once the wxConfig object is destroyed - which should happen automatically on closing the program. Since on the Mac only one instance of wxMaxima is running at a time that should be sufficient. At least in theory.

I am not sure we are talking about the same thing here. There is a settings file (wherever this is on a Mac) and the wxmaxima-init.mac file, which is read by Maxima on startup. The latter must be written before Maxima starts, obviously.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : I had a look at the MacPorts log files (a tricky thing actually ..).

The interesting difference is
```
20.01.2: (works)
:info:build [100%] Linking CXX executable ../wxMaxima.app/Contents/MacOS/wxmaxima
:info:build [100%] Built target wxmaxima

20.01.3: (misses executable)
:info:build [100%] Linking CXX executable wxmaxima.app/Contents/MacOS/wxmaxima
:info:build [100%] Built target wxmaxima
```
If you scroll to `src/CMake_List.txt`in
(https://github.com/wxMaxima-developers/wxmaxima/compare/Version-20.01.2...Version-20.01.3#diff-af3b638bc2a3e6c650974192a53c7291L105) you see relevant changes around line 130).

I must say I don't know much about CMake install specs on Mac. What might be helpful is the observation that the folder from which MacPorts does the installation is in both cases the same:

```
/opt/local/var/macports/build/_Users_msoegtrop_MacPorts_macports-local_math_wxMaxima/wxMaxima/work/destroot/Applications/MacPorts/wxMaxima.app
```

So it looks like the executable is created in the wrong folder.

**Comment by MSoegtropIMC** [comment]
P.S.: the executable ends up after the build in
```
/opt/local/var/macports/build/_Users_msoegtrop_MacPorts_macports-local_math_wxMaxima/wxMaxima/work/wxMaxima-developers-wxmaxima-6e496c7/build/src/wxmaxima.app/Contents/MacOS
```
But I think it should be in
/opt/local/var/macports/build/_Users_msoegtrop_MacPorts_macports-local_math_wxMaxima/wxMaxima/work/wxMaxima-developers-wxmaxima-6e496c7/build/wxmaxima.app/Contents/MacOS
```
The latter folder does exist, but it is empty.

**Comment by gunterkoenigsmann** [comment]
I believe the src/CMakeLists.txt lines that tell cmake where to install the binary are:

~~~~
    install(TARGETS wxmaxima
	     BUNDLE  DESTINATION .
	     RUNTIME DESTINATION bin
	     COMPONENT Runtime)
~~~~

Alternatively it could probably be the
~~~~
    set_target_properties(wxmaxima PROPERTIES
	MACOSX_BUNDLE TRUE
        RUNTIME_OUTPUT_DIRECTORY_DEBUG   ${CMAKE_BINARY_DIR}/wxMaxima.app/Contents/MacOS
        RUNTIME_OUTPUT_DIRECTORY_RELEASE ${CMAKE_BINARY_DIR}/wxMaxima.app/Contents/MacOS
	RESOURCE "${RESOURCE_FILES}"
~~~~
I just re-added the `RUNTIME_OUTPUT_DIRECTORY_DEBUG` and `RUNTIME_OUTPUT_DIRECTORY_RELEASE`. Let's see if it compiles.

**Comment by gunterkoenigsmann** [comment]
I think I found it: You do an out-of-source build. Which means: You create a separate build directory and compile the program from there, which is the recommended method, but as no-one does this it isn't tested and therefore doesn't work. Compiling the thing directly from the top directory of the source tree (typing "cmake ." instead "cmake ..") should resolve that issue.

**Comment by MSoegtropIMC** [comment]
Removing the build folder doesn't fix this. The executable then ends up as
`./src/wxmaxima.app/Contents/MacOS/wxmaxim` while it should be `./wxmaxima.app/Contents/MacOS/wxmaxima`. Should I give your above commits a try? Do you have some favourite documentation for Cmake and Mac?

**Comment by gunterkoenigsmann** [comment]
I hope the commits now work. What they do is to inform cmake where to put the binary - and (on macOS the binary won't work until all .dylibs it used are moved into the right place and are told that they now have to work from there) to inform the 2nd step that moves and relinks the dylibs where to put the binary.

What I don't understand is why travos copies all files into the Right Place: https://travis-ci.org/github/wxMaxima-developers/wxmaxima

Documentation for Cmake and Mac I didn't find. All I can do is trial-and-error (change the code, then look if the output of https://travis-ci.org/github/wxMaxima-developers/wxmaxima looks plausible, then hope that someone verifies it actually runs).

**Comment by MSoegtropIMC** [comment]
I tried your latest patch 838334aa0fd33892fc9543fe0d33f588e4818857. With this all files end up in the same folder `build/src/wxMaxima.app` and if I change the folder from which I copy I get a target folder with executable. Just it is not at the correct place.
What I get is
```
/Applications/MacPorts/wxMaxima.app//Contents/MacOS/wxmaxima.app/Contents/MacOS/wxmaxima
```
what I need is
```
/Applications/MacPorts/wxMaxima.app//Contents/MacOS/wxmaxima
```
When I copy the files where they belong things are fine.

This seems to be an effect of the cmake path settings. The output folder I get is in
```
/opt/local/var/macports/build/_Users_msoegtrop_MacPorts_macports-local_math_wxMaximaTest/wxMaximaTest/work/wxmaxima-838334aa0fd33892fc9543fe0d33f588e4818857/build/src/wxMaxima.app
```
and contains paths like
```
./Contents/MacOS/wxmaxima.app/Contents/MacOS/wxmaxima
```

Let me try to fix this. One question: what was that you wanted to achieve with the CMake changes from 20.01.2 to 20.01.3?

**Comment by gunterkoenigsmann** [comment]
What I tried (and actually succeeded in) was to

- make wxMaxima work with homebrew
- and to make wxMaxima generate a working .dmg file on https://travis-ci.org/github/wxMaxima-developers/wxmaxima

But perhaps I have now understood the problem:

On the mac one can generate a .app bundle or a binary file. The latter only works on a system that has all .dylib files and similar installed in the same place than the system the package was compiled on. 

I guess if you change

    set(MACOSX_BUNDLE TRUE)

to 

    set(MACOSX_BUNDLE FALSE)

you'll get the stand-alone binary you need - but the .dmg no more is created instead.

But as always with MacOS all I can do is to guess; Perhaps a solution that would fix both would be to add a

    file(COPY ${CMAKE_CURRENT_BINARY_DIR}/wxMaxima.app//Contents/MacOS/wxmaxima.app/Contents/MacOS/wxmaxima DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/)

to src/CMakeLists.txt. But there is no guarantee that I am right this time.

**Comment by gunterkoenigsmann** [comment]
With a bit of luck I have now found a solution: By default no Mac OS X bundle is generated (If I finally understood you right you don't need a bundle, but a binary that trusts the .dylib files that can be found on your system to be present in the places they can be found on your system. But there is still a way to request the bundle generator so bundles can be generated, too.

**Comment by MSoegtropIMC** [comment]
Thanks I will try this (might be tomorrow). I think it might make more sense to generate bundles with MacPorts, because it can bundle the dependencies (maxima, gnuplot, ...) with it.

**Comment by gunterkoenigsmann** [comment]
Macports also should have a dependency mechanism: If your package claims to depend on maxima and gnuplot if someone asks macports to install wxMaxima the packages for maxima and gnuplot will automatically installed, as well. You also should be able to say wxMaxima depends on wxWidgets. So in theory there is no need for bundles.

**Comment by MSoegtropIMC** [comment]
I meant that MacPorts allows you (afaik - I didn't try it) to create a stand alone installer (one which doesn't require MacPorts) which does contain the dependencies.

**Comment by MSoegtropIMC** [comment]
I tried your commit 1b004665e041d932fbcbf5cbd7d94f49ec14e0c6 and this does work. I just installed and started it. I will do some more testing, especially around finding gnuplot.

**Comment by MSoegtropIMC** [comment]
I did some testing:

- gnuplot is not found unless I create a startup file with `gnuplot_command:"/opt/local/bin/gnuplot"$` in it
- The issue that wxMaxima puts a bogus gnuplot command in the settings window is gone (it doesn't put anything in there
- Putting the correct (above line) gnuplot command in the settings window does work (it writes the wxmaxima-init.mac file correctly)
- This issue that a bogus Debug Message Pane is obscuring the main window is still there. One has to enable the Debug Message Window. As a workaround it might help to create the panes in a different order, so that the last pane is a more useful pane - e.g. the table of contents window. I can then document that one should keep this pane open.

For gnuplot I will try to come up with a PR.

**Comment by gunterkoenigsmann** [comment]
The idea to create the worksheet as a last pane is brilliant. Thanks a lot! I just committed a fix for that issue, hopefully.

For the gnuplot autodetection: wxMaxima doesn't really need gnuplot - except if it wants to popup embedded plots interactively. But maxima needs gnuplot => I guess the best place to add the gnuplot autodetection would be in maxima (http://maxima.sourceforge.net), not in wxMaxima, because this project needs to be solved for Xmaxima, the guys that use maxima on the console, emaxima, imaxima, maxima-fishbowl etc., too. 

Or it could be addressed in macports: Macports should add the path to gnuplot to the system's search path so "gnuplot" [without any path to it] suffices as program name.

**Comment by gunterkoenigsmann** [comment]
But the idea that if you tell MacPorts, not cMake to create a dmg package we get a maxima bundled with all its dependencies is great. I believe, though, that if you do that additional heuristics for wxMaxima is needed in order to find maxima, and for maxima in order to find gnuplot...

**Comment by tomio-arisaka** [comment]
About the patch file:
We can install MacPorts to a path other than /opt/local.
<https://guide.macports.org/#installing.macports.source.multiple>
So wxMaxima port file must support any paths to installed multiple MacPorts copies.
It is why we need the patch file for wxMaxima.
For example, the binary of Maxima 5.43.0 is installed to "/Applications/Maxima.app/Contents/Resources/opt" with MacPorts. And the included wxMaxima 19.05.7 is installed to "/Applications/Maxima.app/Contents/Resources/opt/Applications".


The following note is the information of wxMaxima installed with MacPorts:
(My wxMaxima is version 19.05.7 with wxWidgets 3.1.2 on macOS 10.13.6)

(1) macOS is based on FreeBSD. So executable files are searched for through $PATH.
When installing MacPorts, ${HOME}/.profile is modified in order to make the MacPorts directory a top priority.

```
$ cat ${HOME}/.profile

# MacPorts Installer addition on 2017-11-22_at_01:02:03: adding an appropriate PATH variable for use with MacPorts.
export PATH="/opt/local/bin:/opt/local/sbin:$PATH"
# Finished adapting your PATH environment variable for use with MacPorts.

$ 
$ echo $PATH
/opt/local/bin:/opt/local/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin
$ 
$ which maxima gnuplot
/opt/local/bin/maxima
/opt/local/bin/gnuplot
$ 
```


(2) When opening wxMaxima on Terminal.app, wxMaxima (maxima) can find gnuplot installed with MacPorts.

Open wxMaxima with the next command on Terminal.app:
`$ open -a /Applications/MacPorts/wxMaxima.app/`

After that, an example on wxMaxima:
```
(%i2)	system ("echo $PATH");
/opt/local/bin:/opt/local/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/usr/local/bin
(%o2)	"#<SB-IMPL::PROCESS :EXITED 0>"

(%i3)	system ("which maxima gnuplot");
/opt/local/bin/maxima
/opt/local/bin/gnuplot
(%o3)	"#<SB-IMPL::PROCESS :EXITED 0>"
```
It means that wxMaxima inherits $PATH from Terminal.app (shell).

(3) On the other hand, when opening wxMaxima on Finder, wxMaxima (maxima) cannot find gnuplot installed with MacPorts.

An example on wxMaxima:
```
(%i2)	system ("echo $PATH");
/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin
(%o2)	"#<SB-IMPL::PROCESS :EXITED 0>"

(%i3)	system ("which maxima gnuplot");
(%o3)	"#<SB-IMPL::PROCESS :EXITED 1>"
```

The $PATH inherited from Finder does not include the MacPorts directory.

(4) After opening wxMaxima, the configuration data is saved to the "${HOME}/Library/Preferences/wxMaxima Preferences" file.

**Comment by MSoegtropIMC** [comment]
@tomio-arisaka : I understand the motivation. But I don't think it is the best way to do it, because it leads to unnecessary maintenance effort since patches break easily. A better way of doing it is to have the patch with the placeholder upstream. One can compile it in only e.g. if OSX_MACPORTS is defined and set this during cmake configure time and use a default otherwise. This way the mechanism won't break by upstream changes.

I did work myself with this kind of patch for years and found that it produces a lot of unnecessary work, at least if one wants to maintain this regularly. Meanwhile I try to avoid this unless absolutely necessary. The most common reason for using this kind of patch is if the upstream does not want it, e.g. if they say they don't care about Mac and don't want Mac specific patches. One can't say that this is the case for wxMaxima.

**Comment by MSoegtropIMC** [comment]
@tomio-arisaka : the solution I came up with is this:

in in wxmaxima/src/Dirstructure.cpp:
```
// This macro is used to mark the macports prefix, so that it can be easily patched
// The MacPorts port file for wxMaxima uses this, so don't remove this!
#define OSX_MACPORTS_PREFIX "/opt/local"
```
and then later e.g.
```
maximaLocation = OSX_MACPORTS_PREFIX "/bin/maxima";
```
In the port file I have:
```
post-patch {
    reinplace -W ${worksrcpath} "s|OSX_MACPORTS_PREFIX  *\"/opt/local\"|OSX_MACPORTS_PREFIX \"${prefix}\" // patched by MacPorts|g" src/Dirstructure.cpp
}
```
The macro construct is safe to search and replace and if it is not patched, say in a local build, it defaults to the default MacPorts folder.

What do you think about this solution?

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann , @tomio-arisaka : I created PR #1233 to fix the gnuplot issue.

@gunterkoenigsmann : the commit 1b004665e041d932fbcbf5cbd7d94f49ec14e0c6 does work for me, but your later commit changing the cmake files doesn't. Does it change the folder in which wxMaxima.app is put?

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : one more note: when I compile with wxWidgets 3.2, I have very serious issues with scrolling. Essentially what happens is that when I scroll the window does not scroll, but instead and starts to get white from the top, until it is completely white. The lower part of the window stays where it is. With wxWidgets 3.0 this behaves as expected.

**Comment by gunterkoenigsmann** [comment]
@tomio-arisaka: I believe that supporting MacPorts is important and that you are right: If the package maintainer cannot test it it will break from time to time.

@MSoegtropIMC/ 1b00466: Actually I don't know if it did. I just removed all lines I felt not to be important. But you have write access to the repo so feel free to do anything that is needed in order to make both the dmg generation on travis and compiling a package for MacPorts work.

**Comment by gunterkoenigsmann** [comment]
Scroll issues most of the time have to do with the following lines of src/Worksheet.cpp:

    #ifdef ANTIALIASSING_DC_NOT_CORRECTLY_SCROLLED
        PrepareDC(antiAliassingDC);
    #endif    

Perhaps we need to PrepareDC the other DC, too; I moved the decision if that needs to be done into the macro ANTIALIASSING_DC_NOT_CORRECTLY_SCROLLED as the name of this macro half-way tells why we are doing this and because it is cleaner as the nested #ifdef command we had before.

**Comment by gunterkoenigsmann** [comment]
According to the wxWidgets documentation we need not to manually call PrepareDC. And on some platforms and wxWidgets version calling it scrolls the drawing context an additional time. Which is about as bad as not scrolling it, at all. The wxWidgets guys are happy to help if you can provide them with a minimal example (like a patched version of the the "scroll" sample the wxWidgets code comes with). But preparing such a thing requires loads of time and for a platform I don't own I cannot even test if it worked => I rarely send them bug reports.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : let's solve one problem at a time. I think using wxWidgets 3.0 is ok for the time being. Let's first get a MacPorts file out and then we can handle the scrolling issues.

Regarding gnuplot: I changed this in my PR so that gnuplot is searched every time. On OSX this doesn't make that much sense - at least right now - because wxMaxima sets `gnuplot_command` in maxima, and what wxMaxima reads back from maxima is then obviously this value. But as far as I can see this is different on other platforms, so it is fine to change this. I agree that as next step we should ensure that maxima gives us a propper gnuplot path and remove the heuristics here.

Thanks for the repo write access! I will create a branch to get the CI and MacPorts builds aligned, but it will likely be Thursday. Do you have a link to the repo etiquette? E.g. are reviews required, is it allowed to merge own changes into master?

**Comment by gunterkoenigsmann** [comment]
No rules: If something breaks I'll try to fix it. And I always commit directly to the repo.
If something that fixes a problem for one platform breaks something in another platform normally the CI finds the solution. And "one thing at a time" sounds fine. If you tell me I can merge the pull request now I'll do that.

**Comment by gunterkoenigsmann** [comment]
Also if you tell me MacPorts now works I will make a new release.

**Comment by MSoegtropIMC** [comment]
> Also if you tell me MacPorts now works I will make a new release.

Not yet with the latest commit - there is still the build issue I get from your commit to fix Travis. I won't have time before Thursday to look into this.

> If you tell me I can merge the pull request now I'll do that.

Yes this PR should work fine. I tested it only on Mac, but I don't expect that it makes a difference on other platforms.

**Comment by tomio-arisaka** [comment]
When installing wxMaxima with MacPorts, most MacPorts users expect the wxMaxima to use Maxima installed with MacPorts.
But if the "/Applications/Maxima.app" file or the "/usr/local/bin/maxima" file exists, Maxima installed with MacPorts is not used as default due to the search sequence of the "Dirstructure::MaximaDefaultLocation" function.
I think the wxMaxima port file needs a patch in order to solve the issue.

**Comment by MSoegtropIMC** [comment]
Indeed. I will change this.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : I am looking into getting to a Mac version which compiles in MacPorts and travis, but I don't see how CI is setup for this project. Do you start travis manually?

**Comment by MSoegtropIMC** [comment]
P.S.: I did set up travis on my fork so that it runs automatically on pushes, but I find this setup a bit unusual. In other projects I work on CI runs automatically on each PR and master branch commit. Maybe it does here as well and I just don't see the links to it.

**Comment by gunterkoenigsmann** [comment]
Theoretically if you create a PR or a branch in the main repository any commit should cause a CI run, in this project, as well. It it doesn't perhaps something in my .travis.yml is wrong.

---

The scrolling problem with newer wxWidgets might be fixed by adding more conditions to the following lines:
````
#if wxCHECK_VERSION(3, 1, 2)
#if wxCHECK_VERSION(3, 1, 3)
#else
#define DC_ALREADY_SCROLLED 1
#endif
#else
#ifdef __WXGTK3__
#else
#endif
#endif
````

**Comment by MSoegtropIMC** [comment]
How do I access the travis information? I don't see any links to CI in the wxMaxima main repo. As I said I enabled travis on my fork but there I also have to login to travis to see the results.

**Comment by gunterkoenigsmann** [comment]
In https://github.com/wxMaxima-developers/wxmaxima/commits/master (and the respective page for each branch) there is a small green hook after the date the commit was made, if both travis and appveyor are happy with the commit - or a red cross, if they aren't. Additionally you can log into travis and tell it under which conditions you want to be sent a mail. I have set this to "status change" which means get a mail about 15 minutes after I have broken something and about 15 minutes after it has been fixed again.

Seems like forks inherit the contents of the repos, but not the webapps; if you want to add a "travis status" and "appveyor status" badge to appear somewhere in the gh_pages branch (that contains the wxMaxima homepage) or if you find a way to configure things in a better way feel free to do so.

**Comment by gunterkoenigsmann** [comment]
Forgot to say: The links to travis and appveyor hide behind the red cross or green hook in the "commits" view on github.

**Comment by MSoegtropIMC** [comment]
Thanks - looks like I should go and get a coffee ;-)
I have the MacPorts build working with latest master now by using the same build commands as travis, notably `-DMACOSX_BUNDLE=TRUE`. I guess it doesn't hurt to keep the build commands consistent.

One question: what do `-DUSE_CPPCHECK=YES`and `-DUSE_OPENMP=YES` do?

My status is:
- latest master builds
- random issues with scrolling with both wxWidgets 3.0 and 3.2. I guess it is better if I nail it to 3.2 and just fix it (following your advice above)
- The debug messages pane must be enabled or everything is messed up
- Sometimes nothing happens when I select a menu command, especially file open. I need to try a few different points in the menu. This was with wxWidgets 3.0. I switched to 3.2 now since the scrolling issues are the same. The scrolling issues  just happen randomly, e.g. sometimes when I rerun Maxima the issues go away.

I think we should agree on a recommended version for wxWidgets. Is this 3.2?

**Comment by MSoegtropIMC** [comment]
One more note: wxWidgets 3.2 has additional scrolling issues to 3.0. Both have random effect that when I am at the bottom I can't scroll up. But 3.2 has in addition the "white slider" issue I described above. I will try to fix this following your suggestion.

One more thing: when I load a file, the `Cell, evaluate all visible cells` menu is grayed out and the keyboad shortcut does not work, but `Cell, evaluate all cells` does work and after running it it fixes the scrolling issue and also that `Cell, evaluate all visible cells` is grayes. Do you also have this with latest master?

**Comment by MSoegtropIMC** [comment]
A question regarding the scroll version hacking: I have this:
```
./var/macports/registry/portfiles/wxWidgets-3.2-3.1.2_0 
```
Does this mean that wxWidgets 3.2 is wxWidgets 3.1.2? I have wxWidgets 3.2 installed, but it seems to point to wxWidgets 3.1.2. Is this normal?

**Comment by gunterkoenigsmann** [comment]
On Linux 3.2 works better than 3.0... ...and 3.2 might be more similar to future versions => if I had a choice I would prefer 3.2. But the white slider bug sounds like a problem we might not be able to work around...
...the grayed out buttons and inability to scroll sounds like something I will be able to find this evening.

For the debug pane issue in wxMaximaFrame::wxMaximaFrame I have tried to make the worksheet the last pane that is instantiated.  If that doesn't work we need a new idea. But everything you write sounds like we make rapid progresses. That is great!

**Comment by MSoegtropIMC** [comment]
Hmm, the wxWidgets page `https://www.wxwidgets.org/downloads/` says 3.1.3 is the latest development release. Why the hack can I install wxWidgets 3.2 on MacPorts? If you could shed some light on this mystery, it would help.

**Comment by gunterkoenigsmann** [comment]
I wondered half of this day how I managed to miss the 3.2 release.

3.1 is the development version becaise it contains one breaking change.

---

Iseem to remember the commands that might resolve the menus and scroll bug might be UpdateMenus and UpdateMLast. But they both in theory should be called automatically by OnIdle...

**Comment by MSoegtropIMC** [comment]
I still don't get it. Is there something like a wxWidgets 3.2 already? I just saw that people are working on this. On MacPorts this is really strange. It says there is 3.0 and 3.2 and 3.2 actually installs 3.1.2.

Since latest stable release appears to be 3.0.4, I think I will stick to that.

**Comment by MSoegtropIMC** [comment]
OK, reading the docs helps. The wxWidgets MacPorts port file contains this:
```
# * 'wxWidgets-3.0'
#   You should generally use this one unless there are valid reasons for
#   a different choice.

# * 'wxWidgets-3.2'
#   At the moment still mainly for development purposes, for testing ports
#   for compatibility with wxWidgets 3.1/3.2.
#   It uses wxWidgets 3.1 at the moment, but will switch to wxWidgets 3.2
#   after the first release of version 3.2 becomes available.
#   If you already want to provide a special variant for your port with
#   'wxWidgtes-3.2', you might only need a revbump after the switch to 3.2
```

**Comment by gunterkoenigsmann** [comment]
Didn't respond to the libomp and cppcheck question: cppcheck scans the c++ source code for common errors and warns when it finds something that might loog fine, but can go wrong => you don't really need it.
openmp (typically needs a libomp install) is a thing that is important, instead: When OpenMP and libomp are detected at build time by cmake wxMaxima can delegate several time-consuming tasks (like scanning the maxima manual for hooks and command names which otherwise halts the program for many seconds) to background tasks that run on other CPUs than the one that runs the wxMaxima GUI.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : 

> For the debug pane issue in wxMaximaFrame::wxMaximaFrame I have tried to make the worksheet the last pane that is instantiated. If that doesn't work we need a new idea.

Can you please provde a commit for this?

The scrolling (and probably also the other issues) more looks like that after loading a file there is something running which always drags the current scroll position to what it was when the file was loaded. This loop is ended by recalculating cells, but otherwise seems to go on for at least a long time. I think the issue that some menu items are not updated properly is the same.

Any idea what this could be? Can I put something into the code to e.g. dump the call tree during scroll events?

**Comment by gunterkoenigsmann** [comment]
The commit should be 2245bd6e398a58a177a62018a0072204b097f932.
...and the unability to scroll IMHO should either be caused by libomp not being installed when cmake configured the project or by the missing commands I will push in the next commit.

**Comment by MSoegtropIMC** [comment]
Thanks! I will check libomp. I set the define and I don't see complaints from cmake, but I don't install it specifically. Not sure if OSX has it be default.

**Comment by gunterkoenigsmann** [comment]
For some reason OSX comes without libomp, even if XCode understands all the #pragma commands that cause calls to libomp, if libomp is installed.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : just wanted to let you know that I am still on the OpenMP stuff. It is indeed not there and it is not sufficient to install the library because it is a compiler feature. One needs rather complicated options. Still experimenting.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : you have a minimum date of 201510 in the CMake file for OpenMP. Is there a good reason for that? The OSX version I get from MacPorts is 201409. Not that it is unreasonable to request a version which is not ancient, but also making a portfile for openmp might get a bit involved.

**Comment by gunterkoenigsmann** [comment]
I wasn't sure if I picked the right minimum openMP date the program compiles with. If it compiles with an older version enabled feel free to correct the minimum date.

On homebrew cmake seems to autodetect which compiler switches are needed. All that needs to be installed manually is libomp-dev. That means: Nearly all of my mac knowledge is based on manual brute-force tests on what compiles and what doesn't. But it at least compiles. The question is: Does macports offer a newer cmake version that automatically guesses the compiler switches?

**Comment by gunterkoenigsmann** [comment]
And don't be afraid: I'm always looking forward to the next release and both working support for the mac and all the improvements the other team members introduce are improvements I am quite proud of. But that a thing, if done right, needs time I know.

---

On ms windows directly after opening a file the scrollbar of the worksheet kind of works in the first attempt: It scrolls once you release it. After that it fully works.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : ok with a lot of cmake arguments and some patching of the minimum version I got it compiled with libomp (it shows in activity monitor that it loads the shared library).

My cmake configure arguments are now:
```
cmake -DwxWidgets_CONFIG_EXECUTABLE:FILEPATH=${wxWidgets.wxconfig} -DUSE_CPPCHECK=YES -DUSE_OPENMP=YES -DMACOSX_BUNDLE=TRUE -DCMAKE_CXX_FLAGS=-I/opt/local/include/libomp -DOpenMP_CXX_FLAGS='-Xpreprocessor -fopenmp' -DOpenMP_CXX_LIB_NAMES=libomp -DOpenMP_libomp_LIBRARY=/opt/local/lib/libomp/libomp.dylib ..
```
As far as I can tell the OpenMP specific defines are the minimum set.

The OpenMP_CXX_SPEC_DATE  version CMake displays is 201107, but I think this is wrong and something like a minimum version CMake displays if it can't figure it out. Anyway, wxMaxima compiles with this.

My cmake is `cmake version 3.16.5`.

I was looking for the cmake definition files which control OpenMP, but couldn't find this. Can you point me to the respective file (on whatever OS - I guess the filename is the same). Or is this all hard coded in CMake?

Ah yes, the test results: enabling OpenMP does not help.

The effect I see is that after loading a file wxMaxima runs at 100% CPU for indefinite time (at least5 Minutes) and insists on setting the scroll position to what it was when I change it. When I srcoll fast with the keyboard it will eventually scroll, but it goes 10x up/down until it goes once in the direction I want. After recomputing all cells it behaves as expected (0% CPU unless it does something, one can scroll). Any ideas what this might be? Points where I could put some logging in? Or is it time to get out the debugger?

**Comment by MSoegtropIMC** [comment]
I just found that OSX has a cool feature: In Activity Monitor one can just select a process and from the toolbar choose "Sample process" and it creates call tree statistics for 3s with 1ms sample interval. The call tree is usually the same. Maybe this tells you something ... I took a sample immediately after start and 7min after start (the one I attached) - it looks the same.

[CallTreeStats.txt](https://github.com/wxMaxima-developers/wxmaxima/files/4419379/CallTreeStats.txt)

**Comment by gunterkoenigsmann** [comment]
How cmake and OpenMP interact I don't really know. But if it compiles now using OpenMP that is a big step forward.

And an integrated gprof? Nice! Let's see if I interpreted the result right. I still don't understand why the font routines are called this open. But at least on my linux computer wxMaxima is now way faster. thanks a lot!

If this doesn't help send me a new CallTreeStats and we'll find the next bottleneck.

**Comment by MSoegtropIMC** [comment]
Yes this integrated profiler is really cool - one can ask average users to do this and it is nice that it easily connects to running jobs, so that one can take spot samples in different situations. Here is a sample *after* I recompute all cells - looks clean.
[CallTreeStats_AfterRecmpute.txt](https://github.com/wxMaxima-developers/wxmaxima/files/4419467/CallTreeStats_AfterRecmpute.txt)

**Comment by MSoegtropIMC** [comment]
I rebased - the trace look different but the issue is still there. I name the traces after the commit hashes to avoid mistakes. Looks like this profile feature brings us forward quickly ...

[68232d928294768a45ac78b7129ffa9278a228a1.txt](https://github.com/wxMaxima-developers/wxmaxima/files/4419575/68232d928294768a45ac78b7129ffa9278a228a1.txt)

**Comment by MSoegtropIMC** [comment]
Profile for your latest commit is on the way ...

**Comment by MSoegtropIMC** [comment]
Btw.: Is my CMake version OK (cmake version 3.16.5.)?

**Comment by MSoegtropIMC** [comment]
Here is the next profile. It is getting better (menues are updated and more responsive) but the scroll issue is still there and it still takes 100% CPU. Any idea why recomputing all cells fixes things?
[9620b5371dec33e4542a1a51a3db720c6bb59bb9.txt.zip](https://github.com/wxMaxima-developers/wxmaxima/files/4419633/9620b5371dec33e4542a1a51a3db720c6bb59bb9.txt.zip)

**Comment by gunterkoenigsmann** [comment]
Which cmake is new enough for MacPorts I unfortunately don't know. ...and in theory 

````
void Worksheet::ScrollToCellIfNeeded()
{
  if(!m_cellPointers.m_scrollToCell)
    return;
  m_cellPointers.m_scrollToCell = false;

  RecalculateIfNeeded();
````
causes a full recalculate. And it is called in the beginning of the trace log => Seems like you found a serious bug in the program somewhere...

**Comment by gunterkoenigsmann** [comment]
Got it: Adjusting the worksheet size causes a re-calculation of all cell sizes, which costs time => it is not done every time some size has changed but only shortly before the idle task decides nothing high-priority is to be done, and therefore we can adjust the worksheet size, recalculate everything and redraw everything as needed.
Unfortunately adjusting the worksheet size didn't mark the job of adjusting the worksheet size as "done" => the idle task exited in order to allow more high-priority stuff to happen and when wxMaxima fell idle again (normally immediately after) started to adjust the worksheet size and exited.

**Comment by gunterkoenigsmann** [comment]
AppVeyor marked the above commit that mentions this issue with a red cross as I made another commit that improves recalculation further directly afterwards and therefore this CI never finished checking if the first of these two commits ever worked.

**Comment by MSoegtropIMC** [comment]
That did it (many positive emticons here)!

wxMaxima is now very responsive from the start. It still occassionally crashes when closing, but I think this shouldn't hold back a release.

I just need to do some more work on the CMake OMP stuff. Right now I set the limit date to 201107 but as I said I think the date reported by CMake is wrong and if we set it like this we might see strange effects on other platforms.

I will also do some tests the next one or two days and then I will prepare the MacPorts PR.

**Comment by gunterkoenigsmann** [comment]
Cool! That would mean that yodur debugging has drastically improved the efficiency of wxMaxima.
If you can convince a debugger to generate a backtrace of the crash on closing: if there is a backtrace resolving a crash normally is astonishingly  easy.

**Comment by MSoegtropIMC** [comment]
> Cool! That would mean that yodur debugging has drastically improved the efficiency of wxMaxima.
> If you can convince a debugger to generate a backtrace of the crash on closing: if there is a backtrace resolving a crash normally is astonishingly easy.

I see what I can do. It is not that reproducible (say 25%) and it usually doesn't happen when I start it in the debugger. Also likely you might get a backtrace of an exception handler. But I could run it e.g. with address sanitizer enabled.

**Comment by MSoegtropIMC** [comment]
Ah and I forgot: the debug messages window is still there. If the debug message window is disabled in the view menu, it sits full width below the main toolbar. And it is only like this if the main toolbar is enabled and the debug message window is disabled.

One thought: I think on Windows Maxima stores the frame layout in a lengthy format in the registry. I guess it is the same on Mac. Do you know in which file this is? Maybe I have a corrupted layout in there and deleting this file fixes it.

**Comment by gunterkoenigsmann** [comment]
I guess the layout information on the mac is stored in a file somewhere. All I know is that the object that stores the configuration knows that the program is named "wxMaxima" and I believe it to be likely that this name ends up in the file name. If you find that file: Can you send me a copy so I can try to break my wxMaxima, too?

...and crashes that only happen if there is no debugger are hard to fix, unfortunately ;-(

**Comment by MSoegtropIMC** [comment]
There is one more issue: the older version of wxMaxima (19.05) have a Windows Menu. On a Mac when you minimize a Window there seems to be no way to get it back except over this Window menu. Most apps also have a window list in the right mouse menu in dock, but wxMaxima doesn't.

And still sometimes it is sluggish - meaning it holds for a second or two. E.g. I type and nothing happens for 2 seconds and then it continues. This shouldn't be a HW problem (i9 processor within 5% of the top and 32 GB RAM and not much else running). I try to get a profile but I am too slow to open activity monitor, selecting wxMaxima, ... . I guess I have to setup a shortcut key for this. Also sometimes if I select a menu item absolutely nothing happens for a few seconds.

But it is usable, which is a lot of progress. Thanks for the fast help so far!

I try to find the settings file.

**Comment by gunterkoenigsmann** [comment]
The window menu issue sounds like there should be a way to tell MacOs which window is a top-level window. But until now I didn't find that way.

...and sluggishness could also happen with CPU load being 0. I once have seen this - but never could find a way to reproduce it.

**Comment by gunterkoenigsmann** [comment]
Hmmm... ...telling wxWidgets that this is a top-level window is easy: Just make the window a wxFrame - which we actually do. So what else can go wrong? Hmmm...

**Comment by MSoegtropIMC** [comment]
> Hmmm... ...telling wxWidgets that this is a top-level window is easy: Just make the window a wxFrame - which we actually do. So what else can go wrong? Hmmm...

It was there in 19.05. Should I bisect when it went away? I could also bisect the debug message frame issue. This was not there in later 18.X versions if I remember right, definitely not in 18.02. But I will first look for a wxmaxima init file.

**Comment by MSoegtropIMC** [comment]
There is one more odd issue - nut sure if it is Mac related: when I select the same font and size for "Text" and some section header, the font in Text is about a factor of 1.5 smaller than in the section. To get about the same size, I have to choose an a factor of 1.5 larger font size in Text. This works as a work around but it is still odd. I also think that "Text" is always bold, even if I don't set it to bold.

**Comment by gunterkoenigsmann** [comment]
Sometimes a bisect is the only way to find out what is going on;

For the text sizes I wonder if I forgot to scale some of the font sizes along with the worksheet...

**Comment by gunterkoenigsmann** [comment]
With a bit of luck the menu issue is resolved now: I now explicitly tell wxWidgets to add its own entries to the main menu.

**Comment by gunterkoenigsmann** [comment]
With a bit of luck the debug messages issue is resolved now, too: After reading in the style I now set many settings that cannot have possibly changed to the values they ought to have after loading the config.

**Comment by gunterkoenigsmann** [comment]
If the GUI still halts sometimes: Might it be waiting for some background thread the debug sidepane anounces to exit? Analyzing the help file can delay opening the right-click menu by a few seconds if both otherwise would collide.

**Comment by MSoegtropIMC** [comment]
I will give your latest commit a try.

The settings file is on OSX in `./Library/Preferences/wxMaxima Preferences`under home. If I delete the file, the layout is really bogus. I have on the left side the debug messages window and on top three *transparent* windows, one with greek letters, one with math chars and one with some commands. So I can see the debug messages through the greek letters ...

But first let me see how your latest commit looks.

**Comment by MSoegtropIMC** [comment]
I tested your latest commit. The debug message issue is still there. I attached two screen shots. One directly after start (with deleted preferences file and tips closed). The second shot is with the side pane pulled out so that you can see better what happens.

Should I switch to wxWidgets 3.1.2 (It didn't help last time I tried, but who knows).

There is still no Window menu - sorry.

<img width="1283" alt="Screenshot 2020-04-03 at 14 51 17" src="https://user-images.githubusercontent.com/7895506/78362991-50041380-75bb-11ea-95ba-b19ee779afc5.png">
![Uploading Screenshot 2020-04-03 at 14.51.54.png…]()

**Comment by gunterkoenigsmann** [comment]
Wow... ...that is really messed up - and really looks like a bug in wxWidgets as it doesn't happen on Linux nor apparently on Windows...

**Comment by gunterkoenigsmann** [comment]
If you enable and re-disable the sidebars: Do they then work as they should? If that is the case I can try to tell the program to do that manually on startup...

**Comment by MSoegtropIMC** [comment]
> If you enable and re-disable the sidebars: Do they then work as they should? If that is the case I can try to tell the program to do that manually on startup...

Yes. Even if I save a configuration (by closing wxMaxima) after disabling all and enabling DebugMessages and TOC it looks fine immediately after startup. Only if I delete the preferences file it looks like that. But if the DebugMessages pane is not enabled, then the view is also messed after startup - usually by having the Debug messages window overlapping the main window in a similar way as the greek later pane overlaps the debug message window in the screen shot, but this varies a bit.

**Comment by gunterkoenigsmann** [comment]
Hope I do it the Right Way: I now enable all sidebars on start-up and then disable the ones that are supposed to be closed. At least I hope I didn't miss one. But for the application menu I have no idea what that should be: According to the wxWidgets documentation it works automatically and there is no way of hindering wxWidgets from doing so on the mac.

I also found out that the problem with the smart quotes only appears with wxWidgets <3.1.1; If you send me both the opening and the closing smart quote I can manually search-and-replace for them, though.

**Comment by MSoegtropIMC** [comment]
It is getting better. With deleted preferences file and TOD closed and the width of the side panel slightly extended it looks like this:

<img width="1288" alt="Screenshot 2020-04-06 at 09 50 18" src="https://user-images.githubusercontent.com/7895506/78535318-3dbcec00-77ec-11ea-9a86-96d264fd15de.png">

So there is still the over/underlay debug messages window. Note that it is overwritten by the greek letters.

I experimented with debugging over the weekend. CMake is a bit odd when it comes to debug builds - if one doesn't specify a build variant it says that it does a debug build, but it doesn't. One has to tell it explicitly.

My main issue is that I can't make the debugger load the symbols for wxWidgets - maybe I need to compile it rather than install it via MacPorts. Likely anyway a good idea.

But if you have an idea what we can do about the debug messages window ...

**Comment by MSoegtropIMC** [comment]
Btw.: I think technically the two screen shots are not that different. I think in both cases it takes the debug messages window as background of the side panel. Just in one case it is the background of the complete side panel and now it is the background of just a piece of the side panel. Maybe this gives you a clue?

**Comment by gunterkoenigsmann** [comment]
Often finding weird bugs in 3rd party frameworks involves many people explaining exactly what they see - and then understanding that speaking with Sherlock Holmes one didn't only see but did observe.

Perhaps the hint with the size was the one that was needed => tried to change the order of a few commands in order to see if that resolves this issue.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : sorry I didn't find time to test this before today. The pane issue is still there. I am debugging this now.

**Comment by gunterkoenigsmann** [comment]
Cool! Thanks! I still have no idea what this pane does do in a different fashion than the others...

...if you can send me a config file from a broken wxMaxima perhaps I can break my wxMaxima that runs on Linux, too, and try to help debugging the problem.

**Comment by MSoegtropIMC** [comment]
I think the issue might be OSX specific. The main difference to other panes is that it is created first and that it is created before the AUI manager gets its settings and that it is hidden. But I think it should work anyway - let me see.

Regarding the crash on close - I got a call stack. The wxSocketBase this pointer is NULL. I think you should check in idle processing if the connection to Maxima is already closed.
```
wxSocketBase::IsConnected() const (/opt/local/Library/Frameworks/wxWidgets.framework/Versions/wxWidgets/3.0/include/wx-3.0/wx/socket.h:135)
wxMaxima::TryToReadDataFromMaxima() (/opt/local/var/macports/build/_Users_msoegtrop_MacPorts_macports-local_math_wxMaximaGitLocal/wxMaximaGitLocal/work/wxMaximaGitLocal-7ce53cdf/src/wxMaxima.cpp:1627)
wxMaxima::OnIdle(wxIdleEvent&) (/opt/local/var/macports/build/_Users_msoegtrop_MacPorts_macports-local_math_wxMaximaGitLocal/wxMaximaGitLocal/work/wxMaximaGitLocal-7ce53cdf/src/wxMaxima.cpp:4541)
wxEvtHandler::ProcessEventIfMatchesId(wxEventTableEntryBase const&, wxEvtHandler*, wxEvent&) (@wxEvtHandler::ProcessEventIfMatchesId(wxEventTableEntryBase const&, wxEvtHandler*, wxEvent&):34)
wxEvtHandler::SearchDynamicEventTable(wxEvent&) (@wxEvtHandler::SearchDynamicEventTable(wxEvent&):31)
wxEvtHandler::TryHereOnly(wxEvent&) (@wxEvtHandler::TryHereOnly(wxEvent&):18)
wxEvtHandler::DoTryChain(wxEvent&) (@wxEvtHandler::DoTryChain(wxEvent&):19)
wxEvtHandler::ProcessEvent(wxEvent&) (@wxEvtHandler::ProcessEvent(wxEvent&):30)
wxEvtHandler::SafelyProcessEvent(wxEvent&) (@wxEvtHandler::SafelyProcessEvent(wxEvent&):9)
wxWindowBase::SendIdleEvents(wxIdleEvent&) (@wxWindowBase::SendIdleEvents(wxIdleEvent&):24)
wxAppBase::ProcessIdle() (@wxAppBase::ProcessIdle():39)
wxApp::ProcessIdle() (@wxApp::ProcessIdle():12)
wxCFEventLoop::CommonModeObserverCallBack(__CFRunLoopObserver*, int) (@wxCFEventLoop::CommonModeObserverCallBack(__CFRunLoopObserver*, int):30)
__CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ (@__CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__:12)
__CFRunLoopDoObservers (@__CFRunLoopDoObservers:123)
__CFRunLoopRun (@__CFRunLoopRun:263)
CFRunLoopRunSpecific (@CFRunLoopRunSpecific:122)
RunCurrentEventLoopInMode (@RunCurrentEventLoopInMode:81)
ReceiveNextEventCommon (@ReceiveNextEventCommon:145)
_BlockUntilNextEventMatchingListInModeWithFilter (@_BlockUntilNextEventMatchingListInModeWithFilter:25)
```

**Comment by MSoegtropIMC** [comment]
P.S.: please don't do any more experimental changes - this makes it harder for me to sync later. Preferably undo the last change.

**Comment by MSoegtropIMC** [comment]
Regarding the log pane: I think the issue is that the Log Window receives a lot of draw messages from the logging between creation and attachment to the AUI manager. Together with being the first child of the frame, this leads to confusion. It can be fixed by either not logging anything or by creating another child first which doesn't receive draw messages before it is attached to the AUI manager. I chose the history windows, but I guess there are better choices. Btw.: not all panes can be created before the log pane. E.g. the main document pane needs to be created after reading the settinge because it tries to connect to Maxima. And reading the settings must be done after creating the log pane because it does logging.

What is your opinion what the least active pane is? Possibly the greek letter pane? I guess it doesn't do anything unless one clicks on it.

**Comment by gunterkoenigsmann** [comment]
History isn't active until commands are executed.
And the worksheet is active, but always visible. Also one can define a variable of the type

WxUpdateablocker wxUpdateBlocker(m_logPane); and it will hinder wxWidgets from sending paint events to that pane as long as this variable exists.

Will look at the back trace as soon as I am at home again: thanks to your help that sounds like an easy fix; if you want to revert things I did: You have write access. Feel free to do so.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : I think the pane issue is solved by creating one of the less active panes before the log pane. I will commit after some more testing.

The only really bad remaining issue is this:
http://trac.wxwidgets.org/ticket/18641
https://forums.wxwidgets.org/viewtopic.php?t=46362&p=195663

Essentially what happens is that you click a menu and nothing happens until something happens. This is about 50% reproducible and clicking around for 3 seconds usually fixes it (I sometimes just click 10 times the same menu). Once it works it seems to work until Maxima exits.

The correspondign keyboard shortcuts do work. Do you have an idea where I could put some logging to possibly track down the root cause?

Switching around Windows helps but not always.

Besides that we have the crash on exit, for which I sent you the trace.

Any idea what we can do about the menu stuff? Can we put a message in a prominent place, say in the main window on start?

**Comment by gunterkoenigsmann** [comment]
The last few entries in thst bug report basically say: If you compile your application to an standalone binary MacOs won't allow it to have a menu. If you compile it to an AppBundle it will have a working menu, instead.
In the beginning before we introduced a way to compile wxMaxima as a standalone binary instead of as an AppBundle I seem to remember you tried to to copy the standalone binary that was contained in the AppBundle somewhere else. If you instead copy the whole AppBundle to that destination and name it wxmaxima instead of wxMaxima.app - would that work?

---

The crash on exit should be resolved in the current git master. At least I hope I didn't wildly misinterpret the backtrace.

**Comment by MSoegtropIMC** [comment]
I am doing nothing special. MacPorts calls cmake, cmake creates an app bundle and MacPorts copies it where it belongs (to folder /Applications/Macports/wxMaxima.app/). I don't think it has to do with that. Also the Menu does work after a while and also in 50% of the cases after start. I see what I can do.

**Comment by gunterkoenigsmann** [comment]
Wasn't it you who wanted to be able to disable the MACOSX_BUNDLE flag? If it wasn't you the current master should build fine and provide working menus again.

**Comment by MSoegtropIMC** [comment]
Regarding MACOSX_BUNDLE: I think this is something you did to get my builds and the Travis build aligned. I am always building app bundles (or at least I think so). As I said the menu issue has nothing to do with that - the issue definitely does happen with app bundles.

But there are good news: it looks like Apple fixed the issue with the recent update to OSX Catalina 10.15.4 (which I installed last night).

So I will pack together my stuff and make a PR for MacPorts over the weekend :-)

**Comment by MSoegtropIMC** [comment]
P.S.: I have `-DMACOSX_BUNDLE=TRUE` in my cmake configure command. I guess I can remove this now?

**Comment by gunterkoenigsmann** [comment]
I think you can remove the -DMACOSX_BUNDLE=TRUE now: Seems like MacOs only likes bundles, anyway.

And this is really good news. as soon as the PR is merged I will make a new release.

**Comment by MSoegtropIMC** [comment]
OK, I am just doing the change requested by @tomio-arisaka to prefer the MacPorts path for maxima in case this is compiled with MacPorts (otherwise it preferes the homebrew path). Then I push it.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : somehow git tells me that I don't have write rights, so I did a PR.

**Comment by gunterkoenigsmann** [comment]
Github says you never accepted the write rights => Did remove and re-add you to the developer team hoping this will cause github to ask again.

**Comment by MSoegtropIMC** [comment]
Yes, sorry for that. You have to accept the invite within a few days and I was busy then. I accepted it now :-) I will do some more tests with the tagged release and then make a PR for MacPorts (likely tomorrow).

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : two friends tested it (which revealed a bug in the macports file).

The PR is out now: https://github.com/macports/macports-ports/pull/6920. I will close this as soon as the PR is accepted.

**Comment by KubaO** [comment]
> openmp (typically needs a libomp install) is a thing that is important, instead: When OpenMP and libomp are detected at build time by cmake wxMaxima can delegate several time-consuming tasks (like scanning the maxima manual for hooks and command names which otherwise halts the program for many seconds) to background tasks that run on other CPUs than the one that runs the wxMaxima GUI.

I wouldn't bother with openmp. It's a non-portable mess and unnecessary these days. Instead, C++ standard library offers everything one needs to execute asynchronous chunks of work. I'll look at it at some point, since Windows does not offer recent-enough openmp and multithreading support would help there. From my perspective, the existing openmp code should be only a stop-gap until it's done properly...

**Comment by gunterkoenigsmann** [comment]
Openmp is practical as it keeps a queue of background jobs that are executed when a CPU is free so we'll never use more concurrent jobs than the CPU offers us.
And the only windows compiler that doesn't support a recent openmp is msvc that normally generates code that is too be slow to be useful. So things aren't too bad right now.

But openmp feels like a dead end to me, too => if you manage to find a better solution that would be great.
I shortly looked into pthreads that is said to be a library one can get running on most systems. But I didn't like adding more dependencies.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : I just wanted to let you know that the new MacPorts file has been merged. One remaining issue is that for people with case sensitive file system (on OSX one can choose, not case insensitive is the default) there are inconsistencies between the names wxMaxima and wxmaxima. E.g. in CMakeList.txt we have:
```
project(wxmaxima LANGUAGES CXX)
set(PACKAGE wxmaxima)
set(APP_NAME wxmaxima)
```
but the final application name in MacPorts is expected to be wxMaxima. What is the correct name? wxmaxima or wxMaxima?

If it is wxmaxima, I will change the MacPorts file accordingly. If it is wxMaxima, we need to discuss how to fix it - either change it consistently in the CMake files or work around it in the MacPorts file.

Btw.: I removed the development version from MacPorts, since it is not GIT master but something arbitrary, which is IMHO completely useless. I see if I can support GIT master, but this is tricky. Anyway, you might want to adjust the notes in (https://wxmaxima-developers.github.io/wxmaxima/download.html).

**Comment by ryandesign** [comment]
> I see if I can support GIT master, but this is tricky.

MacPorts has a policy not to allow ports to be published that install from git master. The portfile author must select a specific commit, verify that it builds, and commit that into the portfile. The portfile author can update this as often as desired.

**Comment by MSoegtropIMC** [comment]
> MacPorts has a policy not to allow ports to be published that install from git master. The portfile author must select a specific commit, verify that it builds, and commit that into the portfile. The portfile author can update this as often as desired.

A mistake IMHO. The `-devel` builds I have seen are mostly useless. It should be clear to everybody that a master branch is without warranty, but it would developers give an easy way to ask users to check if a bug they reported has been fixed by a change in master. I have commit rights to wxMaxima, but not to MacPorts, so I cannot change it as frequently as I want.

**Comment by ryandesign** [comment]
Discussion of MacPorts policy is probably off-topic in this wxMaxima issue. I just wanted to make sure you were aware of the policy before you spend time implementing something that would not be accepted. If you disagree with the policy, the [macports-dev mailing list](https://lists.macports.org/mailman/listinfo/macports-dev) would be a good place to bring up a discussion.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : can you please answer the above question on the preferred name of wxmaxima or wxMaxima? On case sensitive file system this is an important question. Currently the MacPorts works only on OSX with case insensitive setting (which is the default but some people took the more reasonable approach of making it cases sensitive). Depending on what your answer is, the fix would look different.

**Comment by gunterkoenigsmann** [comment]
Didn't see the question. Sorry for that.

I personally prefer all letters in program names to be lower-case: Uppercase letters look too much like someone having an over-eager marketing department and blinking gif buttons everywhere. Well... ...that is my personal opinion. But if you prefer the program name with capital letters that would be fine, too: The important thing is having mac support - and with Tomio, you and Kuba (who does real magic modernizing the code) as new members I believe we can do real great things.

**Comment by ryandesign** [comment]
I would agree that *command line* program names are typically all lowercase. However macOS application names are typically capitalized "properly". (The names of all of the applications Apple ships in macOS have capital letters and spaces where appropriate.)

**Comment by gunterkoenigsmann** [comment]
Then it sounds like in this case using a capital M is the way to go...

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : we are mostly discussing the name of the application folder / application package. Currently cmake produces an application package named "wxmaxima" and I try to copy a package names "wxMaxima" which works on case insensitive OSX only.

Since the question of capitalizing wxMaxima seems to be OS dependent, the question is still if I should change the application name in cmake from "wxmaxima" to "wxMaxima" - which would be for all OSes - or simply change the macports file to rename the application package from wxmaxima to wxMaxima whily copying it to the applications folder.

Or should it be so that the application folder is named wxMaxima but the binary itself is named wxmaxima?

**Comment by gunterkoenigsmann** [comment]
The binary I would keep as lowercase-only as changing the case might break something on Linux and since I often start the program from command-line where I am used to programs being lowercase-only. For the rest: Feel free to change whatever is needed/is more canonical/looks cool.

**Comment by gunterkoenigsmann** [comment]
The binary I would keep as lowercase-only as changing the case might break something on Linux and since I often start the program from command-line where I am used to programs being lowercase-only. For the rest: Feel free to change whatever is needed/is more canonical/looks cool.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : OK. I will then not touch the cmake file and instead simply change the copy operation in the portfile to translate from wxmaxima to wxMaxima for the app folder name.

**Comment by KubaO** [comment]
I guess I'm the one responsible for this, and I consider it high priority, but it simply won't happen this month.

**Comment by MSoegtropIMC** [comment]
> I guess I'm the one responsible for this, and I consider it high priority, but it simply won't happen this month.

Not sure what you mean. The 20.04 version is released on macports and I am currently the official maintainer of the MacPorts build. What has been discussed latest here with @gunterkoenigsmann is not yet implemented because there are much more serious issues. Essentially no wxWidgets app works properly on OSX 10.15 - the menues don't work. This is what is priority on the Mac.

**Comment by KubaO** [comment]
Yeah but 10.15 is not a huge problem since that’s not the only platform the port targets and anyone depending on wxMaxima (like myself) will stay on 10.14 until it gets fixed (or do a trial update, determine what’s broken, and then downgrade back). But according to macports online ports viewer, the port doesn’t build at all. So maybe that’s incorrect and it works? – wouldn’t know, I just build it from sources and don’t bother with that particular port.

Which reminds me that I have to update the maxima port :)

**Comment by MSoegtropIMC** [comment]
> Yeah but 10.15 is not a huge problem

I would think that the majority of OSX users update their OS regularly and 10.15 is meanwhile quite old.

> But according to macports online ports viewer, the port doesn’t build at all

It doesn't build on OSX with case sensitive file system - which is a minority since it is not the default. But since the health status looks bad I should indeed fix it. It is trivial - I didn't do so as yet because I wanted to fix the menue issue, but it looks like this is deeply buried in OSX, so I will keep this for later.

 > Which reminds me that I have to update the maxima port :)

Can you tell me what exactly you want to do? As I mentioned I am currently the maintainer of the port (see https://ports.macports.org/port/wxMaxima/summary) and I would appreciate if you discuss issues with me and we solve this together.

**Comment by MSoegtropIMC** [comment]
@KubaO, @gunterkoenigsmann : I did a few test builds. Neither 20.06.6 nor the latest commit fix the cut and paste issue I reported. (#1309).

Since 20.04 is well tested, I would wait for a fix for #1309 and test then.
Meanwhile I push a fix for (https://trac.macports.org/ticket/60457) based on 20.04. This should also fix the red flags on the MacPorts page.

**Comment by KubaO** [comment]
The fix for that bug will land in 12.07 so that’ll be a good time to step up. The performance difference between 12.06 and 12.07 is let’s say substantial, and I’ve done lots of bug triage to figure out what’s super urgent and what’s less do and what’s not valid any longer.

**Comment by MSoegtropIMC** [comment]
@KubaO : I guess you mean 20.07 - this would be perfect. Maybe I can make progress on the Menu issue on OSX 10.15 meanwhile and then we can make a substantial new release for Mac.

**Comment by KubaO** [comment]
Yes, for reasons unknown I keep substituting 12 for 20 in the version recently 🤪

**Comment by MSoegtropIMC** [comment]
@KubaO : the port health status looks good now:

(https://ports.macports.org/port/wxMaxima/summary)

The older OSes have not been rerun yet - I guess this takes a while.

I will close the issue, since since a while the latest port is 20.04 and not 18.02. Please create a new issue to create a Portfile e.g. for 20.07 as soon as you think it is ready.

**Comment by gunterkoenigsmann** [comment]
I wouldn't be opposed to a 20.07 this weekend: Kuba had sped up things drastically and we resolved many bugs that could lead to wrong, but plausible equations being shown.

**Comment by MSoegtropIMC** [comment]
OK, I will do some test builds today.

**Comment by gunterkoenigsmann** [comment]
Today I found two bugs I want to look into first => I guess the release unfortunately will need a few more days.

**Comment by MSoegtropIMC** [comment]
OK, I will wait then.

**Comment by gunterkoenigsmann** [comment]
The asserts and the recalculation issues I encountered are now resolved => Will make the release tomorrow, if no-one complains.

**Comment by KubaO** [comment]
No complaints from me. I’ve ran 20.04 and 20.06dev side-by-side and my family was quite happy :) They use it for school, in various ways, so it was a bit of a personal quest to improve the familial quality of life :)

**Comment by gunterkoenigsmann** [comment]
Found one regression I still want to fix (#1348).  But I definitely want to publish the speed improvements as fast as possible.
